selenebuild/selenebuild.moon

```diff
@@ -0,0 +1,209 @@
+-- Selene Build Script
+require("lfs")
+
+ERR = 1
+OK = 0
+
+BLOCKLIST_LINES =
+    ["conf.lua"]: {
+        "%s*require%(\"lib.moonscript\"%)[^\n]*",
+        "package%.path = package%.path %.%. \";%./lib/moonscript/%?%.lua\"[^\n]*\n",
+        "package%.path = package%.path %.%. \";%./lib/lulpeg/%?%.lua\"[^\n]*\n",
+    }
+
+class DirectoryTraversal 
+    new: (func) =>
+        -- Instantiate a new DirectoryTraversal that will apply a function to file paths
+        -- recursively
+
+        @func = func
+
+    traverse: (path) =>
+        -- Traverse a directory recursively and apply a function to files found
+
+        traverse_stack = [ "#{path}/#{entry}" for entry in lfs.dir(path) ]
+        seen = {}
+        rv = OK
+
+        while #traverse_stack > 0
+            entry = table.remove(traverse_stack)
+
+            filename = entry\match("([^/]+)$")
+
+            -- Skip builtin traversals
+            if filename == "." or filename == ".."                
+                continue
+
+            -- If we have seen this entry before, skip it
+            if #[e for _, e in ipairs seen when e == entry] != 0
+                continue
+
+
+            table.insert(seen, entry)
+
+            -- Recurse on directories and call the callback on files
+            if lfs.attributes(entry, "mode") == "directory"
+                for sentry in lfs.dir(entry)
+                    table.insert(traverse_stack, "#{entry}/#{sentry}")
+            else
+                    if not @func(entry)
+                        rv = ERR
+
+        return rv
+
+class SeleneBuild
+    new: (src) =>
+        -- Instantiate a new build of the directory
+        -- given in src
+
+        @src = src
+
+    build_file: (path) =>
+        -- Build a .moon file. Called as a callback by the
+        -- directory traversal. Returns OK on success, ERR on failure
+        rv = OK
+        if path\match("^.+%.moon$")
+            fd = io.popen("moonc #{path}")
+            output = fd\read("*a")
+
+            status = fd\close()
+            if not status 
+                print "Error building #{path}"
+                print output
+                rv = ERR
+        return rv
+
+    clean_file: (path) =>
+        -- Clean a .lua file. Called as a callback by the
+        -- directory traversal. Returns OK on success, and cannot fail
+        if path\match("^.+%.lua$")
+            os.remove(path)
+            print "Removed #{path}"
+
+        return OK
+
+    dist_file: (path) =>
+        -- Clean a .moon file. Called as a callback by the
+        -- directory traversal. Returns OK on success, and cannot fail
+
+        if path\match("^.+%.moon$")
+            os.remove(path)
+            print "Removed #{path}"
+
+        return OK
+
+    build: () =>
+        -- Run the build. Returns the status of the build
+
+        traversal = DirectoryTraversal(@build_file)
+        return traversal\traverse(@src)
```

Comment:
just a nitpick, since DirectoryTraversal only contains one method traverse, may as well have it be a single function `traverse(@build_file, @src)`

Reply:
Ah yeah I was going to put more stuff in there and didn't :) will fix.